### PR TITLE
`pod.spec.terminationGracePeriodSeconds` is a negative then convert to 1

### PR DESCRIPTION
**Pull request by wzshiming**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

It's been 4 releases since it was allowed to change from negative to 1, so I think it's time to just set it to the default conversion

#### Which issue(s) this PR fixes:

Continue #98866
xref #103476

#### Does this PR introduce a user-facing change?

```release-note
Pods which have an invalid negative `spec.terminationGracePeriodSeconds` value will be treated as having terminationGracePeriodSeconds of 1
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NONE
```

### CODE REVIEW

File: pkg/apis/core/v1/conversion.go [hunk 1]
```diff
@@ -306,6 +307,9 @@ func Convert_core_PodSpec_To_v1_PodSpec(in *core.PodSpec, out *v1.PodSpec, s con
 		out.HostUsers = in.SecurityContext.HostUsers
 	}
```
Comment (liggitt) [inline pkg/apis/core/v1/conversion.go:309]: This change will affect all objects that contain PodSpec (like deployments, replicasets, etc).

Can we scope this change to only pod objects (move this to Convert_v1_Pod_To_core_Pod / Convert_core_Pod_To_v1_Pod)

For non-pod objects, adding a warning to warningsForPodSpecAndMeta for negative TerminationGracePeriodSeconds (like "negative spec.terminationGracePeriodSeconds values are invalid and will be treated as a value of 1") would avoid breaking those controllers while still surfacing the issue with their request or manifest

Comment (liggitt) [reply pkg/apis/core/v1/conversion.go:309]: Along with those changes, we need test coverage of behavior when creating and updating a pod

**Comment by wzshiming** [comment]
Done

File: pkg/apis/core/v1/conversion.go (hunk 1 above)
Comment (wzshiming) [reply pkg/apis/core/v1/conversion.go]: Done

File: pkg/api/pod/warnings.go [hunk 2]
```diff
@@ -38,12 +38,14 @@ func GetWarningsForPod(ctx context.Context, pod, oldPod *api.Pod) []string {
...
 		oldMeta *metav1.ObjectMeta
+		oldType *metav1.TypeMeta
 	)
 	if oldPod != nil {
 		oldSpec = &oldPod.Spec
 		oldMeta = &oldPod.ObjectMeta
+		oldType = &oldPod.TypeMeta
 	}
-	return warningsForPodSpecAndMeta(nil, &pod.Spec, &pod.ObjectMeta, oldSpec, oldMeta)
+	return warningsForPodSpecAndMeta(nil, &pod.Spec, &pod.ObjectMeta, &pod.TypeMeta, oldSpec, oldMeta, oldType)
```
Comment (liggitt) [inline pkg/api/pod/warnings.go:48]: there's no need to plumb typeMeta, you can issue the warning unconditionally

File: pkg/api/pod/warnings.go [hunk 3]
```diff
@@ -250,5 +252,11 @@ func warningsForPodSpecAndMeta(fieldPath *field.Path, podSpec *api.PodSpec, meta
 		return true
 	})
 
+	// typeMeta is nil for non-pod objects (e.g. pod templates)
+	// warn if the terminationGracePeriodSeconds is negative.
+	if typeMeta == nil &&
+		podSpec.TerminationGracePeriodSeconds != nil && *podSpec.TerminationGracePeriodSeconds < 0 {
+		warnings = append(warnings, fmt.Sprintf("%s: negative value is invalid, must be greater than or equal to 0", fieldPath.Child("spec", "terminationGracePeriodSeconds")))
```
Comment (liggitt) [inline pkg/api/pod/warnings.go:259]:
```suggestion
		warnings = append(warnings, fmt.Sprintf("%s: must be >= 0; negative values are invalid and will be treated as 1", fieldPath.Child("spec", "terminationGracePeriodSeconds")))
```

**Comment by wzshiming** [comment]
Done

**Comment by thockin** [comment]
Won't this potentially cause apply-loops?  If that changed value gets stored, it's not what the user sent, so they could retry and retry and retry..

Admittedly, it's rare to use Pod directly, so limited impact, but I don't think this is a pattern we can follow elsewhere?

**Pull request merged**